stores/sql/mysql/main.go

```diff
@@ -53,8 +58,214 @@ func (b *MainDatabase) CreateMigrationTable(ctx context.Context) error {
+	if err != nil {
+		return fmt.Errorf("failed to prepare statement to insert slice: %w", err)
+	}
+	defer insertSliceStmt.Close()
+
+	for i := range slices {
+		res, err := insertSliceStmt.Exec(ctx,
+			time.Now(),
+			objID,
+			uint(i+1),
+			nil,
+			slabIDs[i],
+			slices[i].Offset,
+			slices[i].Length,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to insert slice: %w", err)
+		} else if n, err := res.RowsAffected(); err != nil {
+			return fmt.Errorf("failed to get rows affected: %w", err)
+		} else if n == 0 {
+			return fmt.Errorf("failed to insert slice: no rows affected")
+		}
+	}
+
+	// insert sectors
+	insertSectorStmt, err := tx.Prepare(ctx, `INSERT INTO sectors (created_at, db_slab_id, slab_index, latest_host, root)
+								VALUES (?, ?, ?, ?, ?) ON DUPLICATE KEY UPDATE latest_host = VALUES(latest_host), id = last_insert_id(id)`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare statement to insert sector: %w", err)
+	}
+	defer insertSectorStmt.Close()
+
+	querySectorSlabIDStmt, err := tx.Prepare(ctx, "SELECT db_slab_id FROM sectors WHERE id = last_insert_id()")
+	if err != nil {
+		return fmt.Errorf("failed to prepare statement to query slab id: %w", err)
+	}
+	defer querySectorSlabIDStmt.Close()
+
+	var sectorIDs []int64
+	for i, ss := range slices {
+		for j := range ss.Shards {
+			var sectorID, slabID int64
+			res, err := insertSectorStmt.Exec(ctx,
+				time.Now(),
+				slabIDs[i],
+				j+1,
+				ssql.PublicKey(ss.Shards[j].LatestHost),
+				ss.Shards[j].Root[:],
+			)
+			if err != nil {
+				return fmt.Errorf("failed to insert sector: %w", err)
+			} else if sectorID, err = res.LastInsertId(); err != nil {
+				return fmt.Errorf("failed to fetch sector id: %w", err)
+			} else if err := querySectorSlabIDStmt.QueryRow(ctx).Scan(&slabID); err != nil {
+				return fmt.Errorf("failed to fetch slab id: %w", err)
+			} else if slabID != slabIDs[i] {
+				return fmt.Errorf("failed to insert sector for slab %v: already exists for slab %v", slabIDs[i], slabID)
+			}
+			sectorIDs = append(sectorIDs, sectorID)
+		}
+	}
+
+	// insert contract <-> sector links
+	insertContractSectorStmt, err := tx.Prepare(ctx, `INSERT INTO contract_sectors (db_sector_id, db_contract_id)
+											VALUES (?, ?) ON DUPLICATE KEY UPDATE db_sector_id = db_sector_id`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare statement to insert contract sector link: %w", err)
+	}
+	defer insertContractSectorStmt.Close()
+
+	sectorIdx := 0
+	for _, ss := range slices {
+		for _, shard := range ss.Shards {
+			for _, fcids := range shard.Contracts {
+				for _, fcid := range fcids {
+					if _, ok := usedContracts[fcid]; ok {
+						_, err := insertContractSectorStmt.Exec(ctx,
+							sectorIDs[sectorIdx],
+							usedContracts[fcid].ID,
+						)
+						if err != nil {
+							return fmt.Errorf("failed to insert contract sector link: %w", err)
+						}
+					} else {
+						tx.log.Warn("missing contract for shard",
```

Comment:
we should probably keep a count and error out at some point

Reply:
I'll think about it. Still trying to figure out if that is an actual problem or not. But from what I can tell it's not happening a lot.